File: packages/world/board/harness/tictactoe.test.ts
// SPDX-License-Identifier: GPL-2.0-or-later

/**
 * Tic-tac-toe's rules (#866 PR-6), against a world that is nothing but named
 * columns — no wasm, no canvas, the same reason `run.test.ts` runs the
 * triangles project against `TestWorld`.
 *
 * The behaviour reproduced here is the pre-rewrite game's, not its Scheme:
 * `git show 81619e5^:krudd/engine/game/tictactoe/rules.scm` and its
 * `tictactoe_test.c` are the specification these tests port from a global
 * board vector and two Scheme functions to two node kinds and a pure helper.
 * Placement, turn alternation, winning, drawing and restart are asserted the
 * same way the C harness asserted them — a sequence of taps and a readout of
 * the board afterwards — just through `Runner` and `TestWorld` instead of
 * `scene_script_call` and a `struct world`.
 *
 * Run by `cargo xtask test-web`, which `cargo xtask check` runs.
 */

import assert from "node:assert/strict";
import test from "node:test";
import type { Board, Project } from "../src/index";
import { DOCUMENT_VERSION, Runner, winnerOf } from "../src/index";
import { TestWorld } from "./world";

/** The frame length every test steps at. Only its sign and non-zero-ness matter. */
const DT = 1 / 60;

/**
 * The board this suite runs: `spawn-grid` in the start lane, `place-mark` in
 * the step lane, an empty paint lane. `size` and every column-name param are
 * left at the kinds' own defaults, which is the configuration a project would
 * actually ship — restating them here would be exactly the noise
 * `triangles.ts`'s module docs warn a fixture that repeats defaults produces.
 */
function project(): Project {
	const root: Board = {
		title: "tic-tac-toe",
		columns: [
			{ name: "position", kind: "vec3" },
			{ name: "mark", kind: "u32" },
			{ name: "won", kind: "u32" },
			{ name: "turn", kind: "u32" },
			{ name: "over", kind: "u32" },
		],
		nodes: [
			{ id: "start", kind: "start", lane: "start", column: 0 },
			{ id: "spawn", kind: "spawn-grid", lane: "start", column: 1 },
			{ id: "step", kind: "step", lane: "step", column: 0 },
			{ id: "place", kind: "place-mark", lane: "step", column: 1 },
			{ id: "paint", kind: "paint", lane: "paint", column: 0 },
		],
		wires: [
			{
				id: "exec-start-spawn",
				from: { node: "start", port: "out" },
				to: { node: "spawn", port: "in" },
				kind: "exec",
			},
			{
				id: "exec-step-place",
				from: { node: "step", port: "out" },
				to: { node: "place", port: "in" },
				kind: "exec",
			},
		],
	};
	return { version: DOCUMENT_VERSION, root: "root", boards: { root } };
}

/**
 * The same board, minus the wire into `spawn-grid` — so `start()` still
 * ensures the four columns (per `Runner.start`'s own docs) but nothing ever
 * spawns, leaving them at zero length. This is the "before anything is
 * spawned" world the empty-world test needs, and it can only be built this
 * way: skipping `runner.start()` entirely would leave the columns not merely
 * empty but *uncreated*, which throws for a different reason than the one
 * under test.
 */
function projectWithNothingSpawned(): Project {
	const withSpawn = project();
	const root = withSpawn.boards.root;
	assert.ok(root !== undefined);
	return {
		...withSpawn,
		boards: {
			root: {
				...root,
				nodes: root.nodes.filter((n) => n.id !== "spawn"),
				wires: root.wires.filter((w) => w.id !== "exec-start-spawn"),
			},
		},
	};
}

/**
 * The world-space centre of cell `cell` (`row * 3 + col`) at the board's own
 * default `size` of 2 — the same arithmetic `spawn-grid` writes into
 * `position` and `pickCell` quantises against, restated here only so a test
 * can name a cell by number and get the point that lands on it.
 */
function cellCentre(cell: number): { x: number; y: number } {
	const size = 2;
	const half = size / 2;
	const step = size / 3;
	const row = Math.floor(cell / 3);
	const col = cell % 3;
	return { x: -half + (col + 0.5) * step, y: half - (row + 0.5) * step };
}

/**
 * A tap at `cell`, in the normalised screen-space `PointerFrame` carries —
 * the inverse of `TestWorld`'s own `worldXFromScreen`/`worldYFromScreen`, at
 * its default camera (half-width and half-height both 2, `VIEW_EXTENT`). A
 * node under test never sees this inversion; it is only how a test states
 * "the finger came down here" in the units a `RunContext` actually hands out.
 */
function tapAt(cell: number): { x: number; y: number; pressed: number } {
	const { x, y } = cellCentre(cell);
	const camera = { halfWidth: 2, halfHeight: 2 };
	return {
		x: (x / camera.halfWidth + 1) / 2,
		y: (1 - y / camera.halfHeight) / 2,
		pressed: 1,
	};
}

/** A screen point that unprojects well outside the grid, pressed or not. */
function tapOutsideTheGrid(): { x: number; y: number; pressed: number } {
	// Screen (0, 0), the viewport's top-left corner, unprojects to world
	// (-2, 2) at the default camera — outside the size-2 grid, which only
	// covers world x and y in (-1, 1).
	return { x: 0, y: 0, pressed: 1 };
}

/** The four columns this game plays on, read fresh — never cached across a tap. */
function state(world: TestWorld) {
	return {
		mark: world.column("mark") as Uint32Array,
		won: world.column("won") as Uint32Array,
		turn: world.column("turn") as Uint32Array,
		over: world.column("over") as Uint32Array,
	};
}

test("spawn-grid spawns the nine cells, row 0 at the top and cell 0 top-left", () => {
	const world = new TestWorld();
	new Runner(project(), world).start();

	assert.equal(world.slotCount, 9, "a 3x3 grid, no more and no fewer");
	const position = world.column("position") as Float32Array;
	// Cell 0's centre, at the default size of 2: top-left quadrant, x < 0
	// and y > 0 — the same convention `pick.ts` and `pick.test.ts` assert
	// against, restated here so a pick and a draw are checked against the
	// same source of truth rather than only against each other.
	assert.ok((position[0] as number) < 0 && (position[1] as number) > 0);
	// Cell 8's centre: bottom-right, x > 0 and y < 0.
	assert.ok((position[24] as number) > 0 && (position[25] as number) < 0);
});

test("running spawn-grid twice spawns the nine cells once, not eighteen", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	runner.start();
	assert.equal(world.slotCount, 9, "a second run must not spawn a second grid");
});

test("placing on an empty cell marks it and passes the turn to the other player", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { mark, turn, over } = state(world);

	runner.frame(DT, tapAt(0));
	assert.equal(mark[0], 1, "X landed on cell 0");
	assert.equal(turn[0], 2, "the turn passed to O");
	assert.equal(over[0], 0, "still playing");

	runner.frame(DT, tapAt(4));
	assert.equal(mark[4], 2, "O landed on cell 4, its own turn");
	assert.equal(turn[0], 1, "and the turn passed back to X");
});

test("tapping an already-marked cell is a complete no-op", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { mark, turn } = state(world);

	runner.frame(DT, tapAt(0));
	assert.equal(mark[0], 1);
	assert.equal(turn[0], 2);

	runner.frame(DT, tapAt(0));
	assert.equal(mark[0], 1, "still X — an occupied cell is not overwritten");
	assert.equal(turn[0], 2, "and the turn did not move either");
});

test("a press outside the grid, and a frame with no press at all, both do nothing", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { mark, turn } = state(world);

	runner.frame(DT, tapOutsideTheGrid());
	assert.deepEqual(Array.from(mark.slice(0, 9)), new Array(9).fill(0));
	assert.equal(turn[0], 0, "untouched — spawn-grid does not write turn either");

	// A frame with the default, un-pressed pointer sample — what every board
	// that never collects one gets, and what a hover without a tap looks
	// like to a node that does.
	runner.frame(DT);
	assert.deepEqual(Array.from(mark.slice(0, 9)), new Array(9).fill(0));
});

test("completing a line ends the round for that player without flipping the turn", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { mark, won, turn, over } = state(world);

	runner.frame(DT, tapAt(0)); // X
	runner.frame(DT, tapAt(3)); // O
	runner.frame(DT, tapAt(1)); // X
	runner.frame(DT, tapAt(4)); // O
	assert.equal(over[0], 0, "not decided until the completing move");

	runner.frame(DT, tapAt(2)); // X completes row 0-1-2
	assert.equal(over[0], 1, "X won");
	assert.deepEqual(
		[won[0], won[1], won[2]],
		[1, 1, 1],
		"the winning line is marked",
	);
	assert.deepEqual(
		[won[3], won[4], won[5], won[6], won[7], won[8]],
		[0, 0, 0, 0, 0, 0],
		"and no other cell is",
	);
	assert.equal(turn[0], 1, "the winning move does not flip the turn");
	assert.equal(mark[2], 1, "the winning cell itself was placed");
});

test("any tap once the round is over restarts to an empty board with X to move", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { mark, won, turn, over } = state(world);

	runner.frame(DT, tapAt(0)); // X
	runner.frame(DT, tapAt(3)); // O
	runner.frame(DT, tapAt(1)); // X
	runner.frame(DT, tapAt(4)); // O
	runner.frame(DT, tapAt(2)); // X wins
	assert.equal(over[0], 1);

	// The restarting tap is consumed by the restart itself — it places
	// nothing, matching the pre-rewrite `ttt-on-selected`, which calls
	// `ttt-restart` instead of `ttt-place-move` once the game is over.
	runner.frame(DT, tapAt(0));
	assert.equal(over[0], 0, "playing again");
	assert.equal(turn[0], 1, "X to move");
	assert.deepEqual(
		Array.from(mark.slice(0, 9)),
		new Array(9).fill(0),
		"the board is genuinely empty, not just reporting as not-over",
	);
	assert.deepEqual(Array.from(won.slice(0, 9)), new Array(9).fill(0));

	// And the board really is empty: X can play cell 0 anew.
	runner.frame(DT, tapAt(0));
	assert.equal(mark[0], 1);
	assert.equal(over[0], 0);
});

test("a full board with no line drawn settles as a draw, only on the move that fills it", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { over } = state(world);

	// X:0 O:1 X:2 O:4 X:3 O:5 X:8 O:6 X:7 ->
	//   X O X
	//   X O O
	//   O X X   (no three-in-a-row for either side)
	const moves = [0, 1, 2, 4, 3, 5, 8, 6, 7];
	for (const cell of moves.slice(0, 8)) {
		runner.frame(DT, tapAt(cell));
	}
	assert.equal(over[0], 0, "not decided until the last cell");

	const last = moves[8];
	assert.ok(last !== undefined);
	runner.frame(DT, tapAt(last));
	assert.equal(over[0], 3, "a draw");
});

test("a move that completes a line and fills the board at once is a win, never a draw", () => {
	const world = new TestWorld();
	const runner = new Runner(project(), world);
	runner.start();
	const { mark, over } = state(world);

	// X takes column 2-5-8; O takes the rest. X's fifth move, at cell 8, is
	// both the ninth placement (the board is then full) and the move that
	// completes the 2-5-8 column — worked out by hand so that no other line
	// completes earlier by coincidence.
	//   O X X
	//   O O X
	//   X O X
	const moves: readonly [number, "X" | "O"][] = [
		[1, "X"],
		[0, "O"],
		[2, "X"],
		[3, "O"],
		[5, "X"],
		[4, "O"],
		[6, "X"],
		[7, "O"],
		[8, "X"],
	];
	for (const [cell] of moves.slice(0, 8)) {
		runner.frame(DT, tapAt(cell));
	}
	assert.equal(over[0], 0, "eight cells filled, one line short of either");
	assert.equal(mark[8], 0, "the ninth cell is still open");

	const finalCell = moves[8]?.[0];
	assert.ok(finalCell !== undefined);
	runner.frame(DT, tapAt(finalCell));
	assert.equal(
		over[0],
		1,
		"a win, even though this move also fills the last empty cell",
	);
});

test("place-mark against a world with nothing spawned yet is a no-op, not a throw", () => {
	const world = new TestWorld();
	const runner = new Runner(projectWithNothingSpawned(), world);
	runner.start();
	assert.equal(world.slotCount, 0, "start ran, but nothing was spawned");

	assert.doesNotThrow(() => {
		runner.frame(DT, tapAt(0));
	});
	assert.equal(world.slotCount, 0, "still nothing to place a mark on");
});

test("winnerOf finds a winner along each of the eight lines individually", () => {
	const lines: readonly (readonly [number, number, number])[] = [
		[0, 1, 2],
		[3, 4, 5],
		[6, 7, 8],
		[0, 3, 6],
		[1, 4, 7],
		[2, 5, 8],
		[0, 4, 8],
		[2, 4, 6],
	];
	for (const line of lines) {
		for (const player of [1, 2] as const) {
			const marks = new Array(9).fill(0);
			for (const i of line) {
				marks[i] = player;
			}
			const result = winnerOf(marks);
			assert.equal(
				result.winner,
				player,
				`line [${line.join(",")}] should give player ${player}`,
			);
			assert.deepEqual(result.line, line);
		}
	}
});

test("winnerOf reports no winner over an empty board or a board with no completed line", () => {
	assert.deepEqual(winnerOf(new Array(9).fill(0)), {
		winner: 0,
		line: undefined,
	});
	// X O X / O O X / O X O — no line either way.
	const drawish = [1, 2, 1, 2, 2, 1, 2, 1, 2];
	const result = winnerOf(drawish);
	assert.equal(result.winner, 0);
	assert.equal(result.line, undefined);
});

File: krudd/engine/game/tictactoe/tictactoe_test.c
/* SPDX-License-Identifier: GPL-2.0-or-later */
/*
 * tictactoe — the game rules end to end against the real image. It boots s7,
 * registers the scene-* primitives, loads the embedded rules, then builds a board
 * of nine named cells and drives clicks the way the plugin's tick does: hand a
 * cell's id to ttt-on-selected (with the world bound, via scene_script_call). The
 * checks cover placement + turns, a row win, a full-board draw, and click-to-
 * restart — all without a GPU or a browser.
 *
 * A placed X is a three-entity composite (mesh-less parent + two bars); a placed
 * O is one entity. So an entity-count delta after a placement tells X from O,
 * which is how the checks confirm the turn alternates.
 */
#include "world.h"
#include "scene.h"
#include "scene_script.h"

#include "script.h"
#include "log.h"

#include "tictactoe_rules_scm.h"

#include <assert.h>
#include <stdio.h>
#include <stdint.h>

/* One world instance reused across the checks; too big for the stack. */
static struct world w;

/*
 * Nine cells named cell-0..cell-8 at ids 0..8 — a full board to play on. No asset
 * catalog is needed: the rules read names, and a spawned mark's mesh path simply
 * resolves to "unbound", which still creates the entities the counts assert on.
 */
static const char *BOARD =
	"(scene t"
	"  (entity (name \"cell-0\")) (entity (name \"cell-1\")) (entity (name \"cell-2\"))"
	"  (entity (name \"cell-3\")) (entity (name \"cell-4\")) (entity (name \"cell-5\"))"
	"  (entity (name \"cell-6\")) (entity (name \"cell-7\")) (entity (name \"cell-8\")))";

/* Click the cell whose entity id is `cell` (id N is cell-N here). */
static int32_t play(int32_t cell)
{
	return scene_script_call(&w, NULL, "ttt-on-selected", cell);
}

/* Poll *ttt-over*: 0 playing, 1 X won, 2 O won, 3 draw. */
static int32_t status(void)
{
	return scene_script_call(&w, NULL, "ttt-status", 0);
}

/* The X and O tallies, unpacked from ttt-score's X*1000 + O packing. */
static int32_t score_x(void)
{
	return scene_script_call(&w, NULL, "ttt-score", 0) / 1000;
}

static int32_t score_o(void)
{
	return scene_script_call(&w, NULL, "ttt-score", 0) % 1000;
}

static void reset_board(void)
{
	world_reset(&w);
	assert(scene_script_build(&w, NULL, BOARD) == 9);
	scene_script_call(&w, NULL, "ttt-reset", 0);
	/* Zero the tally too, so each check starts from a clean scoreboard. */
	scene_script_call(&w, NULL, "ttt-score-reset", 0);
}

/* Placement, occupied-cell rejection, non-cell rejection, and turn alternation. */
static void test_placement_and_turns(void)
{
	uint32_t base;

	reset_board();
	base = w.count;

	assert(play(0) == 1);            /* X on cell 0 */
	assert(w.count == base + 3);     /* X = parent + two bars */
	assert(play(0) == 0);            /* occupied — no-op */
	assert(w.count == base + 3);
	assert(play(4) == 1);            /* O on cell 4 (turn alternated) */
	assert(w.count == base + 4);     /* O = one entity */
	assert(play(99) == 0);           /* id 99 is no live entity — ignored */
	assert(status() == 0);           /* still playing */
}

/* A completed row ends the game for that player, then a click restarts. */
static void test_win_then_restart(void)
{
	reset_board();

	play(0);            /* X */
	play(3);            /* O */
	play(1);            /* X */
	play(4);            /* O */
	assert(status() == 0);
	assert(play(2) == 1);            /* X completes row 0-1-2 */
	assert(status() == 1);           /* X wins */

	/* With the game over, any click restarts: marks cleared, state fresh. */
	assert(play(0) == 1);
	assert(status() == 0);
	/* The board is empty again, so X can play cell 0 anew. */
	assert(play(0) == 1);
	assert(status() == 0);
}

/* A full board with no line is a draw. */
static void test_draw(void)
{
	reset_board();

	/* X:0 O:1 X:2 O:4 X:3 O:5 X:8 O:6 X:7 ->
	 *   X O X
	 *   X O O
	 *   O X X   (no three-in-a-row for either side) */
	play(0); play(1); play(2); play(4); play(3);
	play(5); play(8); play(6);
	assert(status() == 0);           /* not decided until the last cell */
	assert(play(7) == 1);            /* fills the board */
	assert(status() == 3);           /* draw */
}

/*
 * A win strikes the line through and credits the winner; the tally then survives
 * the restart that clears the marks. The winning move spawns X (parent + two bars)
 * plus the one-entity strike bar — four new entities — which is how the strike is
 * observed without a GPU.
 */
static void test_strike_and_score(void)
{
	uint32_t base;

	reset_board();

	play(0);            /* X */
	play(3);            /* O */
	play(1);            /* X */
	play(4);            /* O */

	base = w.count;
	assert(play(2) == 1);            /* X completes row 0-1-2 */
	assert(status() == 1);           /* X wins */
	assert(w.count == base + 4);     /* X (3) + the strike bar (1) */
	assert(score_x() == 1);          /* credited to X */
	assert(score_o() == 0);

	/* A click restarts: marks and the strike are swept, but the tally carries. */
	assert(play(0) == 1);
	assert(status() == 0);
	assert(score_x() == 1);
	assert(score_o() == 0);

	/* A second X win adds to the same tally rather than resetting it. */
	play(0); play(3); play(1); play(4); play(2);
	assert(status() == 1);
	assert(score_x() == 2);
	assert(score_o() == 0);
}

int main(void)
{
	log_init();
	script_init();       /* loads scene_script.scm (scene-build, scene-entity-build) */
	scene_script_init(); /* registers the scene-* host primitives */
	script_eval(TICTACTOE_RULES_SCM);

	test_placement_and_turns();
	test_win_then_restart();
	test_draw();
	test_strike_and_score();

	printf("tictactoe_test: ok\n");
	return 0;
}
